craft_parts/state_manager/state_manager.py

```diff
@@ -0,0 +1,152 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2021 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Part crafter step state management."""
+
+import itertools
+from dataclasses import dataclass
+from typing import Dict, Optional
+
+from craft_parts.steps import Step
+
+from .states import StepState
+
+
+@dataclass(frozen=True)
+class _StateWrapper:
+    """A wrapper for the in-memory StepState class with extra metadata.
+
+    This is a wrapper class for StepState that adds additional metadata
+    such as the update sequence order to the data loaded from a previous
+    lifecycle run. Metadata is read-only to prevent unintentional changes.
+    """
+
+    state: StepState
+    serial: int
+    step_updated: bool = False
+
+    def is_newer_than(self, other: "_StateWrapper"):
+        """Verify if this state is newer than the specified state.
+
+        :param other: The wrapped state to compare this state to.
+        """
+        return self.serial > other.serial
+
+
+class _EphemeralStates:
```

Comment:
The `EphemeralStates` name was the name given in the first design draft and stayed there since then, but `StateDB` is a better name. (`StateManager` is a diferent class and will appear in a future PR)